2018Q3: engineering

smola opened the pull request:
Engineering OKRs for 2018 Q3.

dennwc commented on README.md:
I don't think it's realistic for Q3.

Let's focus on increasing semantic object coverage for drivers (type info and classes are good next milestone) as well as better documentation and user experience in general.

We still have Java/JS which are very popular, so focus on making them work great should have a higher priority, I think. As an option, we can look at bringing C# support instead that might cover some enterprise use cases.

LA will definitely start looking into C/C++, but stating that it will be ready for the end of Q3 is too optimistic even for OKRs.

eiso replied:
Actually, we'll have to prioritize C# (much easier), C & C++ since without it we can't deliver an enterprise product. I would argue that it's the highest priority item for LA, but am fully aware that we are likely not to deliver C/C++ in Q3.

juanjux replied:
C++ is a driver that we could definitely have in Q3 (I have a [half written but severely outdated native driver](https://github.com/bblfsh/cpp-driver)) **IF** we take some pragmatic shortcuts. The most important one is to not try to parse the branching preprocessor directives and macros but instead get the values from the environment (or the defaults) or allow the user to provide values if needed and work over the result of the preprocessed code. And for running this (pragmatic shortcut number 2) this "preprocessor runner" would need access to the full repository, and maybe a docker image with a build environment and the the 500 most popular C++ libraries installed (this still would not warrant 100% of code working, but I think a high % is achievable). 

This (running the preprocessor before parsing the resulting code) is what a guy from Facebook that was in this year's FOSDEM dinner (sorry, forgot the name) told me they do themselves. Once this is done, CDT's parser works pretty well even for incomplete information.

PS: Both the preprocessor values and the set of libraries preinstalled for it and the UAST extraction to work would be easy for an enterprise customer to configure (not so easy for somebody parsing all of github's C++ code for example).

eiso replied:
@mcuadros @smola what do you think of this approach? 

From a product point of view, we can't be without a C/C++ driver for much longer, so I am open to this short cut.

smola replied:
If I understood correctly, access to the full repository is not something that can happen only at the driver side, it would need to review most of babelfish design assumptions.

For the library stuff, I assume we (or customers) can use custom builds of the C/C++ driver for a given environment, as you suggest. I think that would be an acceptable tradeoff.

mcuadros replied:
The approach to follow, it the one used by the rest of the community, the macros are entities of the language, they are not expended or interpreted in any way. Making in this way I think is archivable in Q3

juanjux replied:
The really problematic things are branching directives (#ifdef, #ifndef, etc). #define for constants and macros are not really problematic, trough parsing the macro bodies could be hairy in some cases, but it could be avoided and just tag the usage of them as "function call" which is actually what CDT does by default.

smola commented:
- Removed the C/C++ driver as per previous discussion: https://github.com/src-d/okrs/pull/68#discussion_r198117980
- Clarified Launch/Release wording (cc @bzz)
- Added key result for PGA mainteinance (cc @vmarkovtsev)

vmarkovtsev commented on README.md:
For which projects? Are any of ML included?

mcuadros replied:
Its a platform, to be used by the projects.

smola replied:
Yes, any project that requires it. Implementation of actual benchmarks and regression tests are up to each team.

vmarkovtsev commented on README.md:
Logs for Go or for Python too? CLI guide - do we have one?

bzz replied:
I belive this will be for all the languages that source{d} uses and is about adding a CLI guide, so we will have one, with best practices documented, pushing further the effort that @smola has already started this quarter on:
 - Configuration https://github.com/src-d/guide/pull/214
 - Error handling https://github.com/src-d/guide/pull/213

smola replied:
@vmarkovtsev Go, Python, Scala

We do have some basics in src-d/guide (merged and/or PRs), and some more specifics will come up, modeled after https://github.com/src-d/borges/pull/294

vmarkovtsev commented on README.md:
I would add:

- [ ] Conduct a happiness survey and react to the unveiled issues

I think Máximo mentioned it.

vmarkovtsev replied:
Also

- [ ] Improve the formal process of speaking at a conference 

We have nothing currently. Even a simple checklist would save us much time and nerves, especially to poor Esther who really struggled with my conferencing season. It starts with notifying her once the approval registered in src-d/conferences and ends with submitting the expenses report. It also includes what to check in the booking confirmation because I had to buy tickets at the airport one hour before the flight two times already and that sucked very much. It also includes the Visa knowledge database, e.g. British visas are tricky.

vmarkovtsev replied:
Also

- [ ] Increase the transparency of the events happening at source{d}

Rationale: there are meetups, get togethers, holidays, beaconfellows, etc. at source{d} Many, MANY times people missed them in the calendar. Even if there are calendar records, it is so easy to forget about them. How about creating a Slack bot which will post event notifications a few (1) days before?

vmarkovtsev replied:
Also

- [ ] Resolve the wifi connectivity problems in Madrid office

I walk with my laptop from the kitchen to my work place and it just hangs. It is not only because of Ubuntu :trollface: Anna and Esther confessed that they had the same problems. And of course MEETING ROOMS CONNECTION QUALITY ISSUES.

bzz replied:
>  Resolve the wifi connectivity problems in Madrid office

🙏 🙏 🙏 🙏  and has benefit of beeing a very measurable KR!

smola replied:
Survey added to https://github.com/src-d/okrs/pull/71

Other proposals forwarded to Slack for the right people to take care of.

vmarkovtsev commented on README.md:
Misprint: performant

bzz commented on README.md:
@mcuadros shall all the teams add their support objectives to OKRs as well?

bzz commented on README.md:
A typo? `s/rease/reuse/`.

Also, @smola do you think it's worth to add PGA index creation binary as another example of application, eventually benefiting from such lib?

bzz commented on README.md:
Not sure I can immediately see the difference between last 2 items, but seems very interesting!

@ajnavarro Sorry if that is a stupid question, but do you think it may be worth adding here a little bit of details on what is `New engine interface based on gitbase` to avoid possible confusion?

smola replied:
@bzz I've just clarified.

- "Integration of gitbase with engine" -> "Engine using gitbase as backend"; that is, engine using gitbase to execute queries
- "New engine interface based on gitbase" -> "Change engine schema to match gitbase's"

bzz replied:
Thank you for clarifications! Makes perfect sence

juanjux commented on 2018-Q2.md:
```diff
@@ -0,0 +1,200 @@
+		* [P0] Support tracing logging
+	* [P1] Successfully execute OKRs methodology
+	* [P1] Remove technical friction between multiple teams ([#19](https://github.com/src-d/okrs/issues/19))
+		* [P0] Introduce monthly empathy sessions or other mechanism where different teams get to try other teams' tools
```
Not only try, but also try to contribute something to the code, even if they're small things.

bzz replied:
I'm not sure if you already raised that, but this section/files is only `Q2 2018`  and basically is about the past and for historical purpose. Just FYI

juanjux commented on 2018-Q2.md:
```diff
@@ -0,0 +1,200 @@
+		* [P1] Run on Apache Spark cluster, for full PGA
+		* [P2] Run on Apache Spark cluster, for full world dataset
+	* [P0] Deliver a reference implementation of function-level similarity ([#43](https://github.com/src-d/okrs/issues/43))
+		* [P0] XXX Need proper key results for the R&D
```
Remove this XXX before merging.

smola replied:
These are the old OKRs for Q2 ;)

smola commented:
Merging initial version plus some of the corrections. I will open new PRs for the changes you've been adding for each team.
